packages/next/pages/_document.js

```diff
@@ -201,7 +203,7 @@ export class Head extends Component {
               name="viewport"
               content="width=device-width,minimum-scale=1,initial-scale=1"
             />
-            <link rel="canonical" href={page} />
+            {<link rel="canonical" href={page.split('.amp')[0] + (hasCanonical ? '' : '?amp=1')} />}
```

Comment:
This should just be -- this element can be excluded all together when there's no canonical version AFAIK:
```suggestion
            {hasCanonical && <link rel="canonical" href={page} />}
```